File: 1498-number-of-subsequences-that-satisfy-the-given-sum-condition/1498-number-of-subsequences-that-satisfy-the-given-sum-condition.py
class Solution:
    def numSubseq(self, nums: List[int], target: int) -> int:
        nums.sort()
        n, mod = len(nums), 10**9 + 7
        pows = [1] + [0] * (n - 1)
        l, r, ans = 0, n - 1, 0
        
        for i in range(1, n):
            pows[i] = (pows[i - 1] * 2) % mod
            
        while l <= r:
            if nums[l] + nums[r] > target:
                r -= 1
            else:
                ans = (ans + pows[r - l]) % mod
                l += 1
        return ans

File: 1498-number-of-subsequences-that-satisfy-the-given-sum-condition/1498-number-of-subsequences-that-satisfy-the-given-sum-condition.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var numSubseq = function(nums, target) {
  nums.sort((a, b) => a - b);
  const [n, mod] = [nums.length, 10 ** 9 + 7];
  const pows = new Array(n).fill(0);
  let [l, r, ans] = [0, n - 1, 0];
  pows[0] = 1;
  
  for (let i = 1; i < n; i++) pows[i] = (pows[i - 1] * 2) % mod;
  
  while (l <= r) {
    if (nums[r] + nums[l] > target) r--;
    else ans = (ans + pows[r - l++]) % mod;
  }
  
  return ans;
};